workbench/src/sidebar.c

```diff
@@ -0,0 +1,1152 @@
+/*
+ * Copyright 2017 LarsGit223
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+/*
+ * Code for setup and control of the sidebar.
+ */
+#include <sys/time.h>
+#include <gdk/gdkkeysyms.h>
+#include <string.h>
+
+#ifdef HAVE_CONFIG_H
+	#include "config.h"
+#endif
+#include "wb_globals.h"
+#include <gtkcompat.h>
+
+#include "sidebar.h"
+#include "popup_menu.h"
+#include "utils.h"
+
+enum
+{
+	DATA_ID_UNSET = 0,
+	DATA_ID_WB_BOOKMARK,
+	DATA_ID_PROJECT,
+	DATA_ID_PRJ_BOOKMARK,
+	DATA_ID_DIRECTORY,
+	DATA_ID_NO_DIRS,
+	DATA_ID_FOLDER,
+	DATA_ID_FILE,
+};
+
+enum
+{
+	FILEVIEW_COLUMN_ICON,
+	FILEVIEW_COLUMN_NAME,
+	FILEVIEW_COLUMN_DATA_ID,
+	FILEVIEW_COLUMN_ASSIGNED_DATA_POINTER,
+	FILEVIEW_N_COLUMNS,
+};
+
+typedef enum
+{
+	MATCH_FULL,
+	MATCH_PREFIX,
+	MATCH_PATTERN
+} MatchType;
+
+typedef struct
+{
+	GeanyProject *project;
+	GPtrArray *expanded_paths;
+} ExpandData;
+
+typedef struct SIDEBAR
+{
+    GtkWidget *file_view_vbox;
+    GtkWidget *file_view;
+    GtkTreeStore *file_store;
+    GtkWidget *file_view_label;
+}SIDEBAR;
+static SIDEBAR sidebar = {NULL, NULL, NULL, NULL};
+
+/* Remove all child nodes below parent */
+static void sidebar_remove_children(GtkTreeIter *parent)
+{
+	GtkTreeIter iter;
+	GtkTreeModel *model;
+
+	model = gtk_tree_view_get_model(GTK_TREE_VIEW(sidebar.file_view));
+	if (gtk_tree_model_iter_children (model, &iter, parent))
+	{
+		while (gtk_tree_store_remove (sidebar.file_store, &iter)) {}
+	}
+}
+
+
+/* Create a branch for a folder */
+static void sidebar_create_branch(gint level, const gchar *abs_base_dir, GSList *leaf_list, GtkTreeIter *parent)
+{
+	GSList *dir_list = NULL;
+	GSList *file_list = NULL;
+	GSList *elem;
+	gchar **path_arr;
+
+	foreach_slist (elem, leaf_list)
+	{
+		if (elem->data == NULL)
+		{
+			continue;
```

Comment:
Is this ever passing this point?

Reply:
If I remeber right, yes. There was an error sceanario where I had a crash if NULL was not catched. I think it happened if a branch is empty because of not matching file patterns.